### PR TITLE
Swift 4 + performance improvements

### PR DESCRIPTION
**Pull request by mbrandonw**
### What

We ran into some serious performance problems while using DoctorPretty for snapshot testing, so I wanted to see if there was anything I could do to fix it. I optimized a few small things (like using `reduce(into:)` with mutation instead of the pure version), but the real bottleneck is with the `best` and `fits` functions, which are recursive. I was able to rewrite `fits` as a plain loop, and that improved performance quite a bit, and we'd probably get _huge_ improvements if we could rewrite `best`.

Also, while working on this I needed to update the code to Swift 4...

### CODE REVIEW

File: Package.swift [hunk 1]
```diff
@@ -1,21 +1,16 @@
-// swift-tools-version:3.1
+// swift-tools-version:4.0
 
 import PackageDescription
-import Foundation
-
-// HACK from https://github.com/ReactiveCocoa/ReactiveSwift/blob/master/Package.swift
-var isSwiftPMTest: Bool {
```
Comment (mbrandonw) [inline Package.swift:7]: I dont think this stuff is needed anymore...

File: Sources/DoctorPretty/Doc.swift [hunk 2]
```diff
@@ -35,7 +35,7 @@ public indirect enum Doc {
     }
     
     public static func text(_ str: String) -> Doc {
-        return str == "" ? .empty : ._text(length: str.characters.count, str)
```
Comment (mbrandonw) [inline Sources/DoctorPretty/Doc.swift:38]: just fixing some swift 4 warnings

File: Sources/DoctorPretty/Enclosed.swift [hunk 3]
```diff
@@ -14,9 +14,9 @@ extension BidirectionalCollection where Iterator.Element == Doc, IndexDistance =
     /// Intersperses punctuation inside docs
     public func punctuate(with punctuation: Doc) -> [Doc] {
         if let d = first {
-            return [d] + self.dropFirst().reduce([Doc]()) { t in
-                let (acc, d2) = t
-                return acc <> [punctuation, d2]
+            return [d] + self.dropLast().reduce(into: [Doc]()) { acc, d2 in
```
Comment (mbrandonw) [inline Sources/DoctorPretty/Enclosed.swift]: theoretically `reduce(into:)` will be more performant

File: Sources/DoctorPretty/Extras.swift [hunk 4]
```diff
@@ -9,5 +9,5 @@
 import Foundation
 
 public func spaces(_ i: Int) -> String {
-    return (0..<i).map{ _ in " " }.joined()
+    return String(repeating: " ", count: i)
```
Comment (mbrandonw) [inline Sources/DoctorPretty/Extras.swift:12]: I mean why not!

File: Sources/DoctorPretty/RenderPretty.swift [hunk 5]
```diff
@@ -100,19 +100,35 @@ extension Doc {
...
-                case (_, _, .empty): return true
-                case let (m, w, .char(_, x)):
-                    return fits(prefix: m, w: w-1, doc: x)
-                case let (m, w, .text(length, _, x)):
-                    return fits(prefix: m, w: w-length, doc: x)
-                case (_, _, .line(_, _)): return true
+                var _w = w
+                var _doc = doc
+
+                while true {
```
Comment (mbrandonw) [inline Sources/DoctorPretty/RenderPretty.swift:108]: here's where I unrolled the recursion to just a loop. it ain't pretty, but it works. I was also able to simplify some of the logic inside.

File: Tests/DoctorPrettyTests/DoctorPrettySpec.swift [hunk 6]
```diff
@@ -18,19 +18,19 @@ extension Doc {
...
 extension Doc: Arbitrary {
     public static var arbitrary: Gen<Doc> {
-		let stringDocGen: Gen<Doc> = String.arbitrary.map{ Doc.text($0) }
-		let lineOrEmptyGen: Gen<Doc> = Gen<[Doc]>.fromElements(of: [ Doc.line, Doc.empty ])
-		return Gen<Doc>.frequency([
-		        (3, stringDocGen),
-		        (1, lineOrEmptyGen)
-		    ]).proliferate(withSize: 10)
-	          .map{ $0.cat() }
+        let stringDocGen: Gen<Doc> = String.arbitrary.map{ Doc.text($0) }
```
Comment (mbrandonw) [inline Tests/DoctorPrettyTests/DoctorPrettySpec.swift:27]: just some indentation fixes

File: Tests/DoctorPrettyTests/DoctorPrettyTests.swift [hunk 7]
```diff
@@ -192,6 +192,28 @@ class DoctorPrettyTests: XCTestCase {
             ].joined(separator: "\n"), doc: doc)
     }
 
+    func testLargeDocument() {
+        let doc = (1...65)
```
Comment (mbrandonw) [inline Tests/DoctorPrettyTests/DoctorPrettyTests.swift:196]: so this is my "large document" test. before the `fits` unrolling, I couldn't get this document to layout using only a paltry 30 `"foo"` strings. now we can go higher than 65, but it takes about a second to layout this document.

I'm sure there are bigger wins to be had.

**Comment by mbrandonw** [comment]
There is one failing test:

screenshot 2017-12-27 17 37 43

this test started failing as soon as I upgraded to swift 4, before even doing my performance optimizations. So maybe something that changed with `String`?

File: Package.swift (hunk 1 above)
Comment (bkase) [reply Package.swift:7]: yup

File: Sources/DoctorPretty/Extras.swift (hunk 4 above)
Comment (bkase) [reply Sources/DoctorPretty/Extras.swift:12]: yeah this is better, haha

**Review by bkase — DISMISSED** [review]
Love it

**Comment by bkase** [comment]
I'll see if I can figure out why that test broke (or if we should just remove it)

**Comment by bkase** [comment]
```
diff --git a/.travis.yml b/.travis.yml
index 052e8df..7288aab 100644
--- a/.travis.yml
+++ b/.travis.yml
@@ -10,5 +10,5 @@ install:
   - eval "$(curl -sL https://gist.githubusercontent.com/kylef/5c0475ff02b7c7671d2a/raw/9f442512a46d7a2af7b850d65a7e9bd31edfb09b/swiftenv-install.sh)"
 script:
   - swift build
-  - SWIFTPM_TEST_DoctorPretty=YES swift test
+  - swift test
 
diff --git a/Sources/DoctorPretty/Enclosed.swift b/Sources/DoctorPretty/Enclosed.swift
index 6a95299..994ea55 100644
--- a/Sources/DoctorPretty/Enclosed.swift
+++ b/Sources/DoctorPretty/Enclosed.swift
@@ -14,7 +14,7 @@ extension BidirectionalCollection where Iterator.Element == Doc, IndexDistance =
     /// Intersperses punctuation inside docs
     public func punctuate(with punctuation: Doc) -> [Doc] {
         if let d = first {
-            return [d] + self.dropLast().reduce(into: [Doc]()) { acc, d2 in
+            return [d] + self.dropFirst().reduce(into: [Doc]()) { acc, d2 in
                 acc.append(punctuation)
                 acc.append(d2)
             }
```

Could you make these changes? The dropFirst() bit fixes that broken unit test

File: Sources/DoctorPretty/Enclosed.swift (hunk 3 above)
Comment (bkase) [reply Sources/DoctorPretty/Enclosed.swift]: accidental dropFirst -> dropLast

Comment (mbrandonw) [reply Sources/DoctorPretty/Enclosed.swift]: oh duh! dangit

**Comment by mbrandonw** [comment]
addressed feedback! let's see what travis says!

**Comment by mbrandonw** [comment]
Not sure how to read the travis failure. One of the tests passed, and the other failed: https://travis-ci.org/bkase/DoctorPretty/jobs/322586738

**Comment by mbrandonw** [comment]
@bkase oh just update travis to Xcode 9.2 and it seems to be good now!

**Review by bkase — APPROVED** [review]

**Pull request merged**